src/go/rdctl/pkg/snapshot/manager.go

```diff
@@ -96,41 +95,71 @@ func (manager Manager) ValidateName(name string) error {
 	}
 	for _, currentSnapshot := range currentSnapshots {
 		if currentSnapshot.Name == name {
-			return fmt.Errorf("invalid name %q: %w", name, ErrNameExists)
+			return fmt.Errorf("name %q already exists", name)
 		}
 	}
 	return nil
 }
 
+func (manager *Manager) WriteMetadataFile(snapshot Snapshot) (err error) {
+	snapshotDir := manager.SnapshotDirectory(snapshot)
+	if err = os.MkdirAll(snapshotDir, 0o755); err != nil {
+		return fmt.Errorf("failed to create snapshot directory: %w", err)
+	}
+	metadataPath := filepath.Join(snapshotDir, "metadata.json")
+	metadataFile, err := os.Create(metadataPath)
+	if err != nil {
+		return fmt.Errorf("failed to create metadata file: %w", err)
+	}
+	defer metadataFile.Close()
+	encoder := json.NewEncoder(metadataFile)
+	encoder.SetIndent("", "  ")
+	if err = encoder.Encode(snapshot); err != nil {
+		return fmt.Errorf("failed to write metadata file: %w", err)
+	}
+	return
+}
+
 // Create a new snapshot.
-func (manager Manager) Create(name, description string) (*Snapshot, error) {
+func (manager *Manager) Create(name, description string) (snapshot Snapshot, err error) {
+	// Report on invalid names before locking and shutting down the backend
+	if err = manager.ValidateName(name); err != nil {
+		return
+	}
 	id, err := uuid.NewRandom()
 	if err != nil {
-		return nil, fmt.Errorf("failed to generate ID for snapshot: %w", err)
+		return snapshot, fmt.Errorf("failed to generate ID for snapshot: %w", err)
 	}
-	snapshot := Snapshot{
+	snapshot = Snapshot{
 		Created:     time.Now(),
 		Name:        name,
 		ID:          id.String(),
 		Description: description,
 	}
-
-	// do operations that can fail, rolling back if failure is encountered
-	snapshotDir := filepath.Join(manager.Paths.Snapshots, snapshot.ID)
-	if err := manager.Snapshotter.CreateFiles(snapshot); err != nil {
-		if err2 := os.RemoveAll(snapshotDir); err2 != nil {
-			err = errors.Join(err, fmt.Errorf("failed to delete created snapshot directory: %w", err2))
+	if err = lock.Lock(manager.Paths, "create"); err != nil {
+		return
+	}
+	defer func() {
+		if err != nil {
+			manager.RemoveSnapshotDirectory(snapshot)
 		}
-		return nil, err
+		_ = lock.Unlock(manager.Paths, true)
```

Comment:
`lock.Unlock()` does a few things, many of which are prone to producing errors. I think we should report these errors. That being said, I don't have any good ideas for doing so.